### PR TITLE
[ENH] Add windows_identical to DirectReductionForecaster

### PR DESCRIPTION
**Pull request by hliebert**
#### Reference Issues/PRs
Implements #6545 

#### What does this implement/fix? Explain your changes.

This adds a `windows_identical` parameter to `DirectReductionForecaster`.
 
The class is modified to allow shifted X with non-identical window sizes in `fit`. 

- To achieve this, the previous `fit_concurrent` method is extended to accommodate `X_treatment="shifted"` (using the `shifted_vars_lag` parameter of `ReductionTransformer`).
- Setting `windows_identical=True` and `X_treatment="shifted"` still simplifies to multioutput regression. 
- To allow for the case of `X_treatment="concurrent"` and `windows_identical=True`, the final `X` is sliced by the correct offset to ensure uniform window size.

#### What should a reviewer concentrate their feedback on?

- I think this does what is intended in `fit`, I'm less certain about `predict` as I'm not fully sure what the expectations are for `"concurrent"`. Please make sure both `fit` and `predict` behave as intended.
- I think the names for the methods could be improved, happy for any suggestions.

#### Did you add any tests for the change?

Not yet.

#### Any other comments?

Other things I noticed that should be addressed (although maybe not in this PR).

- `ReductionTransformer` parameters `shifted_vars`, `shifted_vars_lag` and `shifted_vars_freq` are undocumented. `shifted_vars` is never defined.
- `ReductionTransformer` _always_ imputes the missings that arise due to lagging by `window_length`. This is unexpected and potentially undesired. It should allow for these rows to be dropped. I think one way to solve this would be to modify `Imputer` to allow `"dropna"` as a valid imputation strategy (I think there have been issues discussing this).

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://allcontributors.org/docs/en/emoji-key)
- [x] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

### CODE REVIEW

File: sktime/forecasting/compose/_reduce.py
Comment (VascoSch92) [inline sktime/forecasting/compose/_reduce.py:1894]: the docstring is not so clear. In the parenthesis where you explain what `True` does, you have also an `otherwise`, but then you explain what `False` does and it is different from what is done in the `otherwise` above

File: sktime/forecasting/compose/_reduce.py
Comment (VascoSch92) [inline sktime/forecasting/compose/_reduce.py:1913]: one would expect that default value for bool is False :-) not against taht just to understand why it is true

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py:1894]: Yes, I am aware. I used the docstring from `make_reduction` and amended it. The problem is that - from my perspective - `otherwise` is the expected behavior, but not currently implemented (see my comment about imputation in `ReductionTransformer` above). Happy to delete that part though.

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py:1913]: I do think that `False` (=non-identical windows) would be the more sensible default, but @fkiraly suggested in #6545 that the current behavior should remain the default.

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py:1913]: yes, I suppose the more sensible default indeed is `False`, but current behaviour needs to remian default due to not breaking (or changing logic) behind downstream code, see change guide:
https://www.sktime.net/en/latest/developer_guide/deprecation.html#deprecation-policy

We could think about changing it though, but then we have to go through deprecation-change handling.

Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py:1913]: If necessary, we can go through deprecation-change handling. The current class does explicitly warn about interface changes though (`DirectReductionForecaster is experimental, and interfaces may change.`, ll. 1927 ff.).

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py:1913]: I see - we should still try to be nice and raise a warning that is concrete about the change. If we want to change this. What is your opinion? The `windows_identical=True` option is probably going to be more performant.

Btw, do you know whether the general performance issues have improved through other updates, e.g., boilerplate speedups?

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: I've changed the docstring and formatted it to make it more readable.

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: I've made `windows_identical=False` the default and added a warning about the change. Maybe with certain series and short horizons `windows_identical=True` may make sense, but I don't think it is a sensible default. Typically the most recent past carries important information that you'd want to use in forecasting.

Regarding the warnings in general, tbh I'd favor getting rid of the warnings altogether, the current warning is already very annoying and not easily silenced.

Regarding general performance improvements, I can't say. It's always been performant enough for our use cases. I can check but I don't think we noticed any major changes in performance over . (And compared to the old reduction classes it is impossible to say, because a like-for-like comparison is not possible due to bugs and different behavior.)

Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: As noted in the description, another default that is not as expected in my opinion is the fact that missings due to lagging by `window_length` are always imputed. I think there should at least be an option to drop them.

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: I see: how about we remove the "experimental" warning?

Regarding the second warning:

* we need to warn in *advance*, not *after the change*. For that, let's schedule the change for 0.32.0, using the "change default" pattern: https://www.sktime.net/en/stable/developer_guide/deprecation.html#id1
* for warning, we should use `warn` from `sktime.utils.warnings` and make reference from `self`, so users can silence it via `set_config`.

Comment (fkiraly) [reply sktime/forecasting/compose/_reduce.py]: > I think there should at least be an option to drop them.
> This behavior is also different in the old reduction classes and I think should be adjusted (in another PR).

Agreed - you could contribute that?

Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: >     * we need to warn in _advance_, not _after the change_. For that, let's schedule the change for 0.32.0, using the "change default" pattern: https://www.sktime.net/en/stable/developer_guide/deprecation.html#id1

I've changed everything accordingly. Let's leave the experimental warning for now.

> > I think there should at least be an option to drop them.
> > This behavior is also different in the old reduction classes and I think should be adjusted (in another PR).
> 
> Agreed - you could contribute that?

I can try to have a look at it some time. I would appreciate some guidance on where this would be best implemented (e.g. whether to adjust `ReductionTransformer` or modify `Imputer` directly - are there any side effects to modifying `Imputer`?)

**Comment by fkiraly** [comment]
Re failures - instead of ` offset = -y_lags[lag] - 1`, do you maybe want ` offset = lag - 1` or `offset = lag`, because `lag` already is the element of `y_lags`?

**Comment by hliebert** [comment]
> Re failures - instead of ` offset = -y_lags[lag] - 1`, do you maybe want ` offset = lag - 1` or `offset = lag`, because `lag` already is the element of `y_lags`?

No, because the offset has to be smaller for larger horizons. `lag` here corresponds the lead (negative lag) for the current horizon. The alternative way to determine this is to calculate `max(fh) -1`.

I think the test failures were due to the problem that the default to preserve the current behavior is `windows_identical=True` for `X_treatment="shifted"` and `windows_identical=False` for `X_treatment="concurrent"`. Those should be fixed now.

The remaining failures are due to the fact that the proposed change deprecation procedure actually violates the tests.

```python
FAILED sktime/tests/test_all_estimators.py::TestAllObjects::test_constructor[DirectReductionForecaster] - AssertionError: windows_identical
assert True == 'changing_value'
 +  where 'changing_value' = <Parameter "windows_identical='changing_value'">.default
FAILED sktime/tests/test_all_estimators.py::TestAllObjects::test_set_params_sklearn[DirectReductionForecaster] - AssertionError: get_params result of DirectReductionForecaster (x) does not match what was passed to set_params (y). Reason for discrepancy: [windows_identical].type, x.type = <class 'bool'> != y.type = <class 'str'>
assert False
```

**Comment by fkiraly** [comment]
I'm confused, is `lag` not already the lead lag at that point?

In the failing example, `y_lags = [-5, -2]` (coming from `fh = [2, 5]`), and `lag` is `-5`, so you are trying to do `[-5, -2][-5]`

**Comment by fkiraly** [comment]
> The remaining failures are due to the fact that the proposed change deprecation procedure actually violates the tests.

No, it doesn't 😃 

if you read the procedure and the accompanying template carefully, it explicitly recommend sthat you do not overwrite `self.windows_identical` from the passed value, but use `self._windows_identical` instead, see https://www.sktime.net/en/stable/developer_guide/deprecation.html#examples-to-illustrate-recipes

I admit that it is not explicitly talked about in the example (although the general pattern does), so maybe we (or you?) would like to add some notes to the exapmle? The `_` underscore is easy to miss.

**Review by fkiraly — CHANGES_REQUESTED** [review]
almost done, we should make sure we fix the attribute overwrite

File: sktime/forecasting/compose/_reduce.py
Comment (fkiraly) [inline sktime/forecasting/compose/_reduce.py:1895]: I think the bullet point shoud be with space and one indent less.

That is,

```
Specifies whether

* ` True` : etc
```

**Comment by hliebert** [comment]
> > The remaining failures are due to the fact that the proposed change deprecation procedure actually violates the tests.
> 
> No, it doesn't 😃

Sorry, I missed that, fixed now.

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: Done.

**Comment by hliebert** [comment]
> I'm confused, is `lag` not already the lead lag at that point?
> 
> In the failing example, `y_lags = [-5, -2]` (coming from `fh = [2, 5]`), and `lag` is `-5`, so you are trying to do `[-5, -2][-5]`

It is. I'm not sure I follow you. In this example, the offset for horizon 5/lag -5 should be 0, the offset for horizon 2 should be 3.

File: sktime/forecasting/compose/_reduce.py
Comment (hliebert) [reply sktime/forecasting/compose/_reduce.py]: > I see: how about we remove the "experimental" warning? I could be convinced if you would like to add yourself to the `maintainers` tag of the class, so it has someone who looks after it...
> 
> Regarding the second warning:
> 
>     * we need to warn in _advance_, not _after the change_. For that, let's schedule the change for 0.32.0, using the "change default" pattern: https://www.sktime.net/en/stable/developer_guide/deprecation.html#id1
> 
>     * for warning, we should use `warn` from `sktime.utils.warnings` and make reference from `self`, so users can silence it via `set_config`.

I removed the warning. Since it is thrown in the constructor, it can't be silenced via `set_config`.

**Review by fkiraly — CHANGES_REQUESTED** [review]
I will review in more in-depth - can you add the new parameter to `get_test_params` so the CI runs all the tests?

**Comment by hliebert** [comment]
> I will review in more in-depth - can you add the new parameter to `get_test_params` so the CI runs all the tests?

Done. Thanks for reviewing!

**Comment by fkiraly** [comment]
Great! The tests are failing, because somewhere we end up with zero samples passed to the `sklearn` regressor,
you can use `check_estimator` for easy debugging

**Comment by hliebert** [comment]
I fixed the offset calculation for the in-sample case, all tests are passing now.

**Review by fkiraly — DISMISSED** [review]
Brillant, looks great!

Also requesting a review from @benheid since this is such a complex code location, and he had looked at the reducers recently.

**Comment by hliebert** [comment]
I've adjusted changed the code s.t. `shifted` always uses `_X` in `predict`. O/w future data may be used when available, e.g. during CV.

**Comment by hliebert** [comment]
The tests using windows are failing (though it appears for unrelated reasons).

**Comment by fkiraly** [comment]
Probably one of the soft dependencies causing issues - generally, it is recommended to update from `main` every time you push to the PR, to ensure you do not accidentally pick up failures that are already fixed. Just did that, let´s see what the tests will do.

**Review by fkiraly — APPROVED** [review]
still approve - if tests pass

**Comment by fkiraly** [comment]
> I've adjusted changed the code s.t. `shifted` always uses `_X` in `predict`. O/w future data may be used when available, (during CV). Besides this, we have been using this variant (`X_treatment="shifted", windows_identical=False`) and it works as expected.

Can you also quickly explain why this is fine? Are we not losing data that we should be using in other scenarios? For CV, what is the special case that you are worried about? The interfaces should ensure "no leakage", I thought.

**Comment by hliebert** [comment]
During CV, `X_test` is generated by concatenating the lags (which are generated from `_y`) and other `X` features after transformations. If other features are based on `X` instead of `_X`, there are undesired index values added during concatenation. This potentially leads to a failure due to non-monotonous index values (missing past and future index values are added at the end). I don't see any harm to using `_X` - with `X_treatment=shifted`, I expect to use the same X in train and test.

**Comment by hliebert** [comment]
Ok, after investigating a bit more closely, the cause of this may be an interface issue elsewhere. 

Below is a minimal example that triggers the problem when using `X` instead of `_X` in `predict`. Reproducible using commit `f006b315b05d7bb7636d9b04f51a08ceb380540f`, after also fixing `msg_return_dict="list"` to `"dict"` [here](https://github.com/sktime/sktime/blob/6397e87070b12f8feea9142ace74920371dba5e2/sktime/transformations/base.py#L1307).

```python
import numpy as np
from sklearn.ensemble import RandomForestRegressor
from sktime.datasets import load_longley
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import ForecastingPipeline, TransformedTargetForecaster
from sktime.transformations.series.boxcox import LogTransformer
from sktime.transformations.series.difference import Differencer
from sktime.transformations.series.subset import IndexSubset

from prices_mlheadline.model.reduce import DirectReductionForecaster

y, X = load_longley()

horizon = ForecastingHorizon(np.arange(1, 3), is_relative=True)

reduction_forecaster = DirectReductionForecaster(
    RandomForestRegressor(),
    window_length=3,
    X_treatment="shifted",
    windows_identical=False,
)

# this is needed to align X to y (differencing causes loss of first obs)
pipe_x = ForecastingPipeline(
    [
        IndexSubset(index_treatment="remove"),
        reduction_forecaster,
    ]
)

pipe = TransformedTargetForecaster(
    [
        LogTransformer(),
        Differencer(lags=1, na_handling="drop_na"),
        pipe_x,
    ]
)

pipe.fit(y, X, fh=horizon)
pipe.predict(fh=horizon, X=X)
```

**Comment by fkiraly** [comment]
augh, this seems to be an unpleasant combination case. Can you kindly report this as a bug issue, and the conditions needed to reproduce it assuming we mer this PR as-is?

**Comment by hliebert** [comment]
Thanks, I'll report the transformation issue and open a PR to fix the msg_return_dict issue.

**Pull request merged**
